tools/validate/validate.cpp

```diff
@@ -98,25 +92,48 @@ void scan_file(const filesystem::path& filepath, const TabPolicy tab_policy, vec
                 // [0x20, 0x7E] are the printable characters, including the space character.
                 // https://en.wikipedia.org/wiki/ASCII#Printable_characters
                 ++disallowed_characters;
-                constexpr size_t MaxErrorsForDisallowedCharacters = 10;
-                if (disallowed_characters <= MaxErrorsForDisallowedCharacters) {
-                    fwprintf(stderr, L"Validation failed: %ls contains disallowed character 0x%02X.\n",
-                        filepath.c_str(), static_cast<unsigned int>(ch));
+
+                if (disallowed_characters <= MaxErrorsForErrorLinesReported) {
+                    fwprintf(stderr, L"Validation failed: %ls contains disallowed character 0x%02X at line %u.\n",
+                        filepath.c_str(), static_cast<unsigned int>(ch), static_cast<unsigned int>(line));
                 }
             }
 
             if (ch == CR || ch == LF) {
                 if (prev == ' ' || prev == '\t') {
                     ++trailing_whitespace_lines;
+
+                    if (overlength_lines <= MaxErrorsForErrorLinesReported) {
+                        fwprintf(stderr, L"Validation failed: %ls contains traling whitespace at line %u.\n",
+                            filepath.c_str(), static_cast<unsigned int>(line));
+                    }
                 }
 
                 if (columns > max_line_length) {
+                    if (overlength_lines < MaxErrorsForErrorLinesReported) {
+                        overlength_line_occurecnes[overlength_lines] = line;
+                    }
                     ++overlength_lines;
                 }
                 columns = 0;
             } else {
                 ++columns;
             }
+
+            if (prev == CR) {
+                if (ch == LF) {
+                    has_crlf = true;
+                } else {
+                    has_cr = true;
+                }
+                ++line;
+            } else {
+                if (ch == LF) {
+                    has_lf = true;
+                    ++line;
+                }
+            }

```

Comment:
I know this is existing but

```suggestion
            } else if (ch == LF) {
                has_lf = true;
                ++line;
            }
```